### PR TITLE
ci: add upgrade to provider example test

### PR DESCRIPTION
**Pull request by elchead**
### Context

This extends #2745 by testing cluster upgrades with the provider.

### Proposed change(s)

- add new upgrade section to the e2e test
- refactor `e2e/internal/upgrade` to reuse the upgrade assertion from the e2e upgrade test in the provider
- adjust the upgrade test assert to skip the target image assertion when the value is empty

### Additional info

- [AB#3693](https://dev.azure.com/Edgeless/ae37573d-ccde-4af2-ab1e-001e587197d1/_workitems/edit/3693)

How tested:
```
gh workflow run e2e-test-provider-example.yml -r feat/ci/upgrade-provider-example -f cloudProvider=gcp -f providerVersion=v2.14.2 -f image=v2.14.2 -f toImage=ref/main/stream/nightly/v2.15.0-pre.0.20231229110237-07c884b94501 -f toKubernetes=v1.28.5
```

- ✅ [E2E run on GCP](https://github.com/edgelesssys/constellation/actions/runs/7359410699/job/20034139195)
- ✅ [E2E run on AWS](https://github.com/edgelesssys/constellation/actions/runs/7359412499/job/20034143664)
- ✅ [E2E run on Azure](https://github.com/edgelesssys/constellation/actions/runs/7394026401/job/20114784858)

### Checklist

- [ ] Update [docs](https://github.com/edgelesssys/constellation/tree/main/docs)
- [x] Add labels (e.g., for changelog category)
- [x] Is PR title adequate for changelog?
- [x] Link to Milestone

### CODE REVIEW

**Comment by 3u13r** [comment]
Please fix the PR description as well as the commit history. 
* There should be no blank bullet point in the PR description 
* For me the "proposed changes" is a _must_ category to fill out
* It would be great if commits were prefixed with a category
* There should be no "fix up", 2x "fix upgrade" commits when opening the PR
* The Checklist has checkboxes to be filled out before or right after opening the PR

**Comment by elchead** [comment]
@derpsteb what was the reason to make the status run in go routine? This is problematic because the go routine might still run after the test has failed.

https://github.com/edgelesssys/constellation/actions/runs/7356364135/job/20026389166#step:27:186
```
--- FAIL: TestUpgradeSuccessful (10861.72s)
    upgrade.go:120: 
        	Error Trace:	e2e/internal/upgrade/upgrade.go:120
        	            				src/runtime/asm_amd64.s:1650
        	Error:      	Should be false
        	Test:       	TestUpgradeSuccessful
        	Messages:   	expected at least 5 nodes, got 4
    upgrade.go:114: 
        	Error Trace:	e2e/internal/upgrade/upgrade.go:114
        	            				e2e/internal/upgrade/upgrade.go:58
        	            				e2e/provider-upgrade/upgrade_test.go:66
        	Error:      	Condition never satisfied
        	Test:       	TestUpgradeSuccessful
K8s config path empty. Using environment variable KUBECONFIG=/home/runner/work/constellation/constellation/cluster/constellation-admin.conf.
stdout: Target versions:
```

Also the test reruns even when it passes. I added prints to show this behavior:
```
ivj6skkHxuzcxbk1xv6ZGxrteJxVH7KlX7YRdZ6eARKwLe4	    AFZEAwoKCQ==	    -----END CERTIFICATE-----
Status is available.
Microservices are upgraded.
Nodes are upgraded.
2023/12/29 16:10:09 Node status (2023-12-29 16:10:09.665193 +0000 UTC m=+71.211386042):
2023/12/29 16:10:09 Setting image version: ref/main/stream/nightly/v2.15.0-pre.0.20231229110237-07c884b94501
2023/12/29 16:10:09 Setting K8s version: v1.28.4
2023/12/29 16:10:09 Setting microservice version: v2.15.0-pre.0.20231229125039-57f8888b02cf
```

Any idea why this happens?

**Comment by elchead** [comment]
For the upgrade test to work, https://github.com/edgelesssys/constellation/pull/2788 is required

File: .github/workflows/e2e-test-provider-example.yml
Comment (elchead) [inline .github/workflows/e2e-test-provider-example.yml:326]: As suggested by @3u13r, I'm working on a PR to enforce to set the microservice version explicitly. cc: @msanft

**Comment by elchead** [comment]
Executing a new [run](https://github.com/edgelesssys/constellation/actions/runs/7412423993/job/20169181751) after rebase

**Comment by msanft** [comment]
Postponing this for now, until general upgrades are back to working

**Comment by elchead** [comment]
> Postponing this for now, until general upgrades are back to working

This test is not affected, and all tests are green. The test only has 1 control plane and 1 worker.

**Comment by thomasten** [comment]
@msanft @daniel-weisse Not sure if you both or just one of you should review this.

File: .github/workflows/e2e-test-provider-example.yml
Comment (daniel-weisse) [inline .github/workflows/e2e-test-provider-example.yml:234]: Is it intended for this test to only run on 1 worker + 1 control-plane node?
Seems to be easily changeable to a configurable input

File: .github/workflows/e2e-test-provider-example.yml
Comment (daniel-weisse) [inline .github/workflows/e2e-test-provider-example.yml:222]: nit: the whitespaces in this generated Terraform file are very inconsistent (they were already before this PR)
e.g.
```terraform
terraform {
  required_providers {
    constellation = {
      source  = "edgelesssys/constellation"
      version            = "${version}"
    }
  }
}
```

File: .github/workflows/e2e-test-provider-example.yml
Comment (daniel-weisse) [inline .github/workflows/e2e-test-provider-example.yml:332]: I think this breaks the "upgrade to latest main if not set" behavior
The conditional should be removed, since we either have `inputs.toImage` as non empty, or `steps.find-latest-image.outputs.image`
(Need to double check the template, though I think it should be correct to assume it picks the second, if `inputs.toImage` is an empty string)
```suggestion
          cat >> _override.tf <<EOF
          locals {
            image_version = "${{ inputs.toImage || steps.find-latest-image.outputs.image }}"
          }
          EOF
```

File: .github/workflows/e2e-test-provider-example.yml
Comment (daniel-weisse) [inline .github/workflows/e2e-test-provider-example.yml:306]: We should move this to just before the `Update cluster configuration` step to avoid lingering clusters if building the provider fails.
We may still run into problems if updating the cluster configuration itself fails, but I think thats acceptable for now

**Review by daniel-weisse — COMMENTED** [review]
Only the workflow part for now

File: .github/workflows/e2e-test-provider-example.yml
Comment (msanft) [reply .github/workflows/e2e-test-provider-example.yml:222]: We could easily address this by running terraform fmt ... after writing the file.

File: .github/workflows/e2e-test-provider-example.yml
Comment (msanft) [inline .github/workflows/e2e-test-provider-example.yml:323]:
```suggestion
        working-directory: ${{ github.workspace }}/cluster
        if: inputs.toImage != "" && inputs.toKubernetes != ""
```
nit: Would save us some of the bash below

File: .github/workflows/e2e-test-provider-example.yml
Comment (msanft) [inline .github/workflows/e2e-test-provider-example.yml:399]:
```suggestion
```

File: .github/workflows/e2e-test-provider-example.yml
Comment (msanft) [inline .github/workflows/e2e-test-provider-example.yml:392]: This shouldn't be the case

File: .github/workflows/e2e-test-provider-example.yml
Comment (daniel-weisse) [reply .github/workflows/e2e-test-provider-example.yml:222]: This was more about the formatting in GitHub actions.
I don't think there is a need to have our e2e actions use a file that has pretty formatting.
But having consistent whitespaces helps with reading the test and editing it in the future.

File: .github/workflows/e2e-test-provider-example.yml
Comment (elchead) [reply .github/workflows/e2e-test-provider-example.yml:306]: what do you mean? There is no way that this step will fail due to missing Terraform, because Terraform was executed in the step before.

File: .github/workflows/e2e-test-provider-example.yml
Comment (elchead) [reply .github/workflows/e2e-test-provider-example.yml:323]: it should be possible to only update one of them.

File: .github/workflows/e2e-test-provider-example.yml
Comment (elchead) [reply .github/workflows/e2e-test-provider-example.yml]: good catch

File: .github/workflows/e2e-test-provider-example.yml
Comment (daniel-weisse) [reply .github/workflows/e2e-test-provider-example.yml:234]: Not suggesting to change the default, just thought we might want to leave it configurable if that can be done easily (e.g. we havent hit the inputs limit on the workflow yet).

If not, just ignore.

File: e2e/internal/upgrade/upgrade.go
Comment (daniel-weisse) [inline e2e/internal/upgrade/upgrade.go:71]:
```suggestion
	testMicroservicesEventuallyHaveVersion(t, targetVersions.Microservices, timeout)
	log.Println("Microservices are upgraded.")

	testNodesEventuallyHaveVersion(t, k, targetVersions, wantControl+wantWorker, timeout)
	log.Println("Nodes are upgraded.")
```
nit: the rest of the functions seem to use `log.PrintX` instead of `fmt.PrintX`

Ignore if this was done intentionally for some formatting purposes

File: .github/workflows/e2e-test-provider-example.yml
Comment (elchead) [reply .github/workflows/e2e-test-provider-example.yml:390]: what do you mean? i tested this a few weeks ago and it was complaining in the upgrade test when the cfg was put in the same directory as the binary.

File: .github/workflows/e2e-test-provider-example.yml
Comment (msanft) [reply .github/workflows/e2e-test-provider-example.yml:323]: Then replace the && with a ||😉

Comment (elchead) [reply .github/workflows/e2e-test-provider-example.yml:323]: see comment from Daniel (new code). So the if condition would be incorrect.

**Review by msanft — COMMENTED** [review]
Skimmed over the CI and code changes, which look alright to me. Regarding the logical relations of the upgrade, I can't say really much.

File: e2e/provider-upgrade/upgrade_test.go
Comment (daniel-weisse) [inline e2e/provider-upgrade/upgrade_test.go:35]:
```suggestion
	wantWorker  = flag.Int("want-worker", 1, "Number of wanted worker nodes.")
	wantControl = flag.Int("want-control", 1, "Number of wanted control nodes.")
```
very nit picky: setting 1 as the expected default seems more reasonable for the test, since we expect at least 1 node to upgrade for both control-plane and worker

**Review by daniel-weisse — COMMENTED** [review]
Go code looks good to me

**Comment by elchead** [comment]
@msanft / @daniel-weisse  thanks for the review! But one of you needs to approve to merge this :D I will start a new E2E run before merging

File: .github/workflows/e2e-test-provider-example.yml
Comment (daniel-weisse) [reply .github/workflows/e2e-test-provider-example.yml:306]: Not relevant anymore
When I made the comment there were builds steps after this cleanup which could fail without having a dedicated cleanup of the cluster.

**Review by daniel-weisse — APPROVED** [review]
Good to merge from my side if tests pass

**Pull request merged**
